### ISSUE TITLE
Some methods definitions have got ''out" and "in" switched

### ISSUE THREAD

**Issue by ipaleka** [DanEdens/Virtual_Desktops_Plugin#1]
Hi, I'm here following this [Stackoverflow comment](https://stackoverflow.com/questions/57149456/how-to-implement-windows-10-ivirtualdesktopmanager-interface-in-python#comment100916370_57149456).

Brilliant work, thank you very much for this, I'm working on a GPLv3 project and I'm planning to use your code refactored into my code, I guess linking from code to this project would fit?

This issue is created as some of your methods definition have some "out" and "in" switched, like `IVirtualDesktopManagerInternal.GetDesktops` or object array `IObjectArray.GetCount`.

Also, HRESULT and POINTER are ctypes members, not ctypes.wintypes.

Thank you again, very much!

**Comment by kdschlosser** [comment]
Typically when a pointer object is passed to an interface method ans is to be "filled" with data the direction is "out" it is not a literal meaning it is going to return it to you.

HRESULT and POINTER are ctypes.wintypes in python 2.7

This is not finished.. and it is still being tested It does not work in it's present form. I do not have Windows 10 installed (nor will I ever) This is for a user for another piece of software I work on.

Take a look at this project I have been tinkering with over the past year.. this is not complete yet and it is something that can be used as a reference. It is a verry large portion of the Windows 10 SDK ported to Python.

https://github.com/kdschlosser/pyWinAPI

**Comment by ipaleka** [comment]
> Typically when a pointer object is passed to an interface method ans is to be "filled" with data the direction is "out" it is not a literal meaning it is going to return it to you.

Yes, I understand the point behind `byref` in ctypes, but this COMMETHOD in comtypes implementation kind of doesn't imply *in* if *out* is set and so error is raised if you pass an argument. Changing the `["out"]` to `["in"]` completely solves the problem as byref carries the value back and it can be used afterward.

> HRESULT and POINTER are ctypes.wintypes in python 2.7

That was my first tought, but then I saw that `print(id)` call. :)

> This is not finished.. and it is still being tested It does not work in it's present form. I do not have Windows 10 installed (nor will I ever) This is for a user for another piece of software I work on.

I'm working on a multiplatform GPL project and this is the last piece of API I've missed for Windows. GNU/Linux part was finished weeks ago.

I managed to implement all the needed functionalities, the only one I'm missing is `SwitchDesktop`, probably something about restype being set to HRESULT and in one moment it worked with c_void_p, but then it continued to argue that argument doesn't have QueryInterface method.

Well, dunno exact phrase in English, but in Croatian people say that "morning is smarter than evening". xD

I'll take a look at your project tommorow, that's for sure!

**Comment by kdschlosser** [comment]
> Well, dunno exact phrase in English, but in Croatian people say that "morning is smarter than evening". xD

> I'll take a look at your project tommorow, that's for sure!

I got ya on that one.

And the print() was because of habit now that I have been writing for python 3.7. took me forever to get used to the ()'s and now it's the other way around.. LOL


I have no way of testing so if you post any errors I may be able to help you out. This is difficult because of the lack of Windows API documentation dealing with it. I am happy that at least it gives you a "foot in the door" so to speak.

**Comment by ipaleka** [comment]
Hi!

That pyWinAPI is a tremendous job, do you have any prediction when it can be available for testing as a Python package and which parts of API that alpha/beta may contain?

In my project there are three different approaches to dealing with Windows API: 

1) ctypes.WinDLL calls to user32, kernel32, dwmapi, ...
2) pywin32 calls like GetWindowPlacement, IsIconic, ...
3) this one with comtypes

I have had some plans to refactor all the pywin32 calls to ctypes.WinDLL calls in the future, but if you think your project may be used instead, then that can be the way to uniform things like I kind of managed to do in GNU/Linux with Wnck in PyGObject (with a little bit of python-xlib) and in Mac OS X (pyobjc-framework-Cocoa and pyobjc-framework-Quartz).

I need a specific set of functionalities/API calls, and I'm not worried if your project is incomplete if the domain logic is robust and applicable for the different Windows versions.

And about `SwitchDesktop` problem I have got. First of all, let me explain "in" and "out" problem. If I leave *out*:
```
        COMMETHOD(
            [helpstring("Method GetDesktops")],
            ctypes.HRESULT,
            "GetDesktops",
            (["out"], ctypes.POINTER(ctypes.POINTER(IObjectArray)), "ppDesktops"),
        ),
```
then the `TypeError` is raised:
```
(vd) c:\dev\examples>python virtualdesktop.py
Traceback (most recent call last):
  File "virtualdesktop.py", line 507, in <module>
    pDesktopManagerInternal.GetDesktops(ctypes.byref(pObjectArray))
TypeError: call takes exactly 1 arguments (2 given)
```
So I changed to `["in"]` and desktop's uids are listed:
```
(vd) c:\dev\examples>python virtualdesktop.py
{F7A15DF7-3C90-4327-B702-FCAB45BC7797}
{18DB210D-60D1-4631-9418-D4DA038CFC45}
Traceback (most recent call last):
  File "virtualdesktop.py", line 523, in <module>
    pDesktopManagerInternal.SwitchDesktop(ctypes.byref(pDesktop))
ctypes.ArgumentError: argument 1: <class 'AttributeError'>: 'CArgObject' object has no attribute 'QueryInterface'
```
And there's that error I've got after I declared `pDesktop` as:
```
pDesktop = ctypes.POINTER(IVirtualDesktop)()
```
The arrising of that problem is obvious to predict as I'm sending a reference to pointer just the way you did in a call to `pObjectArray.GetAt`, but there's definition of pointer of pointer (`ctypes.POINTER(ctypes.wintypes.LPVOID)`) and in `SwitchDesktop` that's just a pointer to a class (`ctypes.POINTER(IVirtualDesktop)`). 

Can you tell me which instance should I pass?

P.S. I forgot to ask, do I need to *release* any of the instantiated object, like `pDesktop` for example? I think I read in comtypes docs iirc that it is done automatically for the classes having `IUnknown` as base class, is that right?

**Comment by kdschlosser** [comment]
the problem is if you don't construct the pointer then you cannot use byref on.

it is saying that IVirtualDesktop does not have the QueryInterface method. which is rather strange because it does have it.

try this..

instead of 

```python
pDesktop = ctypes.POINTER(IVirtualDesktop)()
pDesktopManagerInternal.SwitchDesktop(ctypes.byref(pDesktop))
```

change it to
```python
pDesktop = IVirtualDesktop()
pDesktopManagerInternal.SwitchDesktop(ctypes.pointer(pDesktop))
```

**Comment by kdschlosser** [comment]
and yes the pyWinAPI is a very large undertaking.

I have written a code converter that does a pretty decent job at most of it. I do have to do cleanup after I am done. the issue is It's to damned large for any IDE i have tried on the thing. They all pitch a fit. IDK maybe it doesn't like a 1/4 of a million lines of code.

my issue is keeping track of all of the damned imports. and handling the preprocessor macros so the right bits of code get run for the right operating system version and architecture

I started to go cross eyed after about 6 months of working on it. so i had to take a break from it.

the issue with pywin32 is it's buggy. I have had to on more then one occasion monkey patch the thing for bugs that were reported over a decade ago. it seems as tho development has once again started. The biggest problem I have with it is the inconsistent API, missing constants and also the fact that it does not follow what the Windows API is and the documentation is fantastic at telling you what parameters gets passed to a function but it does not tell you where some of those parameters come from LOL

I used it to create a Windows service. That took me a few weeks to get right because there is no documentation on it.

**Comment by ipaleka** [comment]
>it is saying that IVirtualDesktop does not have the QueryInterface method. which is rather strange because it does have it.

Argh, sorry man I have taken your time for this, it was so silly mistake by me that I'm ashamed while I'm explaining: I tried to switch to that just initialized pointer without pointing it to an array item or getting the actual instance its uid. :/

So frustrating, but that's not going to spoil the enjoyment of the moment of having that last piece of the puzzle called Microsoft API. Now the Windows version of the project has pretty much the same functionality as GNU/Linux version and I'm almost ready for the initial public release. :)

**Comment by ipaleka** [comment]
I hope you'll find the motivation to finish pyWinAPI. I've subscribed to watch it here on Github and I hope I'll recognize the point when it maybe can be used in my project.

**Comment by ipaleka** [comment]
And this issue should be closed as wontfix or something as the title is a perfect example to comment with simple RTFM:

https://pythonhosted.org/comtypes/#calling-methods

> Arguments marked [out] or [out, retval] in the IDL are returned from a sucessful method call, in a tuple if there is more than one. If no [out] or [out, retval] arguments are present, the HRESULT returned by the method call is returned. When [out] or [out, retval] arguments are returned from a sucessful call, the HRESULT value is lost.

**Comment by kdschlosser** [comment]
> Argh, sorry man I have taken your time for this, it was so silly mistake by me that I'm ashamed while I'm explaining: I tried to switch to that just initialized pointer without pointing it to an array item or getting the actual instance its uid. :/

give me a code example of the correct way. This way I will be able to use it.


as far as the whole in and out thing is concerned.... the GetDesktops is correct???

```python
pDesktopManagerInternal.GetDesktops(ctypes.byref(pObjectArray))
```

except it should look like

```python
pDesktops = pDesktopManagerInternal.GetDesktops(ctypes.byref(pObjectArray))
```

is what you are saying?

**Comment by ipaleka** [comment]
Actually, it's just:
```
pObjectArray = pDesktopManagerInternal.GetDesktops()
```
and then:
```
count = pObjectArray.GetCount()
```
where count is plain `int`.

The only thing I couldn't do with your signature is `pObjectArray.GetAt`, I changed that `["out", "iid_is"]` to `["in", "iid_is"]` and called it with:
```
pDesktop = POINTER(IVirtualDesktop)()
pObjectArray.GetAt(index, IID_IVirtualDesktop, ctypes.byref(desktop))
return pDesktop.GetID()
```
where that `index` is an int.

Btw, that doubts you have got in comments for `MoveViewToDesktop`, it's for moving windows from other processes to a different virtual desktop. But, of course it's not that simple, that view needs to be referenced by `GetViewForHwnd` method from `IApplicationViewCollection`. The main problem is, afaik for now, that `IApplicationViewCollection` has different uid based on Windows 10 builds. :(

P.S. https://stackoverflow.com/a/55542867/11703358

**Comment by kdschlosser** [comment]
I figured that out as well.

do me a favor I am going to open my own repository for this whole thing, I want to make it into a python library. any of the things I may have incorrect in it do you mind doing a PR wiht the corrections,

I do not have a Windows 10 machine to test with and I do not know if the latest Windows server supports Virtual Dektops. I guss I can start up an appveyor VM and RDP into it and test the thing.

**Comment by ipaleka** [comment]
I suppose you'll make the example calls to api just the way you did at the end of that \_\_int\_\_.py file. I may test that, that's not a problem, but not sure what to do if that `eg` import has to be involved.

Also, I may try to create the example calls if you explain for which methods you need them.

**Comment by kdschlosser** [comment]
No NO that is for a program I develop called EventGhost. Eventually this is going to end up being added as a plugin. But I like to make a separate python library first. then add the library to the GUI related components of the plugin.

When I make the library all of the ctypes interfaces will be wrapped

The entry point for the user will be classes with properties and methods for the user.

Something along the lines of 

```python
import pyWinVirtualDesktop

for desktop in pyWinVirtualDesktop:
    # do code here for desktop.
    for application in desktop:
        # do code here for application.
        pass
```

I believe that the desktop ID's are persistent so they will remain between reboots. I might code in naming the desktops along with any other user settable information and save it to an xml file.

**Comment by ipaleka** [comment]
I managed to finish that IApplicationViewCollection MoveViewToDesktop, together with code and tests for all attributes, methos definition and usage either as private or as public methods to `.windows.vdi.WindowsDesktopsWin10` created for that purpose.

I'll post the link to code with your ® notice in it, coming inside of release v0.3alpha of the software I'v been working on.

**Comment by kdschlosser** [comment]
https://github.com/kdschlosser/pyWinVirtualDesktop

**Comment by kdschlosser** [comment]
The repo this issue is on is a buddy of mine. you can view my original work on this forum.

http://www.eventghost.net/forum/viewtopic.php?p=53308#p53308

I dunno why he made this repo but he did. and I just happen to be poking about and stumbled upon the stackoverflow and this repo.
and that was right about the time my buddy posted a link to this repo on the forum.

If you have modified the code a lot just a simple comment line will do. something like this.

```python
# based on the work by @kdschlosser <link to my repo> <http://www.eventghost.net> <link to that forum post>
```

**Comment by MadTinker** [comment]
I was trying to figure this magic out by making it work, and modifying the script into something closer to a plugin. Going for the dive-in approach.  I didn't really intend it as a public thing. Sorry Ivica, That;s my bad. I've  gotten less time for it than I thought i would, but even still I saw just how much bigger of a project it truly was when you opened yours.. It's some seriously impressive stuff.

**Comment by kdschlosser** [comment]
@DanEdens 

No Worries m8.. it doesn't bother me at all that you made the repo. I am glad you did, it has already helped someone out. 

I am not able to test it because of my lack of Windows 10.

I have expanded upon it a lot thus far. I am right in the middle of adding in all of the "View" crap. I am not sure what an Application View is or what exactly it does but from what I am guessing is it is some kind of a wrapper around an application window to provide a redesigned look and feel. after this addition you will be able to access these views and the following will be available from a view

* suppression of system overlays (get, set)
* visible bounds (get), x, y, width and height
* visible bounds mode (get, set), UseVisible or UseCoreWindow
* full screen system overlay mode (get, set), Standard or Minimal
* full screen (get, set)
* resize (set) width, height
* min size (set) width, height
* application view mode (get, set) Default or Compact Overlay
* exit view
* orientation (get), landscape or portrait
* is on lock screen (get)
* screen capture enabled (get, set)
* title text (get, set)
* title bar colors
    
    *  foreground
    * background
    * inactive foreground
    * inactive background

* title bar button colors

    * foreground
    * background
    * inactive foreground
    * inactive background
    * hover foreground
    * hover background
    * pressed foreground
    * pressed background

**Comment by ipaleka** [comment]
We're all glad for this project. I'm preparing my project for the initial public release and I'll link the module that uses this code refactored.

Btw, I made the unit tests for the classes definition and their methods, should I make a fork of pyWinVirtualDesktop, implement those tests for your classes and then make a pull request?

**Comment by kdschlosser** [comment]
That would be great! Do as many PR's as you like. I know there are bound to be errors as i am working blind on this project. LOL

I will set up an appveyor account using the latest Windows server maybe it has virtual desktop support.

**Comment by ipaleka** [comment]
Here it is, [my implementation](https://github.com/ipaleka/arrangeit/blob/master/arrangeit/windows/vdi.py).

And there are [the tests](https://github.com/ipaleka/arrangeit/blob/master/tests/test_windows_vdi.py).

Those tests don't implement any kind of "magic", I'm just using the documentation for the [COMMETHOD](https://github.com/enthought/comtypes/blob/master/comtypes/__init__.py) and checking if a class has instantiated the correct tuple at the correct position in `_methods_` list.

For the test purposes and as a proposal to you, I'm planning to create the Api class and override its `__getattr__` method. The point is that when somebody instantiates you vdi/api class like 

```
self.vdi = pywinvirtualdesktop.Api()
```
then all the root level methods should be available in the following format:

```
array = self.vdi._get_desktops()
```
and all the methods should have been implemented for the "instantiated" interface objects like:

```
desktop = array._get_at()
```

A "magic" can be involved too. :) Another proposal is that if you change the name of a method from "Method SomeAction" to "Public method SomeAction" then it can be called with:

```
value = self.vdi.some_action(args)
```

**Comment by kdschlosser** [comment]
how did you solve the Class not registered error when using CoCreateInstance on the VirtualDesktopManager

**Comment by ipaleka** [comment]
> how did you solve the Class not registered error when using CoCreateInstance on the VirtualDesktopManager

I'm not sure if I had had that problem, it's working as expected (afaik... :-/) with being set as a class attribute:
```
def _get_manager(self):
    """Instantiates and returns pointer to interface documented by Microsoft.

    :returns: pointer to :class:`IVirtualDesktopManager`
    """
    return comtypes.CoCreateInstance(
        CLSID_VirtualDesktopManager, IVirtualDesktopManager
    )
```
and I call its methods like:
```
desktop_id  = self.manager.GetWindowDesktopId(hwnd)
#
visible = self.manager.IsWindowOnCurrentVirtualDesktop(hwnd)
#
ret_val = self.manager.MoveWindowToDesktop(hwnd, ctypes.byref(desktop_id))
```

**Comment by kdschlosser** [comment]
yeah I got it working.. dunno what I did to get it working i do not believe i changed anything... who knows. I am also using an unknown build of Windows Server 2016 so I am not sure what is and is not supported.

**Comment by MadTinker** [comment]
I have at least 1 windows 10 laptop sitting idle, Tomorrow I'll see if I cant get it setup for you to test on.

**Comment by kdschlosser** [comment]
well I am at a standstill once again. Appveyor is using a 3 year old build of Windows Server 2016. so most of the API for the Virtual Desktops does not exist.

I just sent in a support request to them.

**Comment by ipaleka** [comment]
I'm trying to collect tests after I installed `pytest` and `comtypes`.
```
(pywvd) C:\dev\pyWinVirtualDesktop>python -m pip install pytest comtypes
```
```
Python 3.7.4 (tags/v3.7.4:e09359112e, Jul  8 2019, 20:34:20) [MSC v.1916 64 bit (AMD64)] on win32
Windows 10
  Version: 1809
  OS build: 17763.107
```

<details>
  <summary>Output of collecting tests error</summary>

```
(pywvd) C:\dev\pyWinVirtualDesktop>python -m pytest -v
================================================= test session starts =================================================
platform win32 -- Python 3.7.4, pytest-5.0.1, py-1.8.0, pluggy-0.12.0 -- C:\dev\venvs\pywvd\Scripts\python.exe
cachedir: .pytest_cache
rootdir: C:\dev\pyWinVirtualDesktop
collecting ... Windows fatal exception: access violation

Current thread 0x00000de8 (most recent call first):
  File "C:\dev\venvs\pywvd\lib\site-packages\comtypes\__init__.py", line 1172 in Release
  File "C:\dev\venvs\pywvd\lib\site-packages\comtypes\__init__.py", line 918 in __del__
  File "C:\Users\Gama Info\AppData\Local\Programs\Python\Python37\lib\inspect.py", line 732 in getmodule
  File "C:\Users\Gama Info\AppData\Local\Programs\Python\Python37\lib\inspect.py", line 780 in findsource
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\_code\source.py", line 236 in findsource
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\_code\code.py", line 63 in fullsource
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\_code\code.py", line 205 in getsource
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\_code\code.py", line 584 in _getentrysource
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\_code\code.py", line 661 in repr_traceback_entry
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\_code\code.py", line 715 in repr_traceback
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\_code\code.py", line 765 in repr_excinfo
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\_code\code.py", line 535 in getrepr
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\nodes.py", line 280 in _repr_failure_py
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\nodes.py", line 330 in repr_failure
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\runner.py", line 261 in pytest_make_collect_report
  File "C:\dev\venvs\pywvd\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "C:\dev\venvs\pywvd\lib\site-packages\pluggy\manager.py", line 81 in <lambda>
  File "C:\dev\venvs\pywvd\lib\site-packages\pluggy\manager.py", line 87 in _hookexec
  File "C:\dev\venvs\pywvd\lib\site-packages\pluggy\hooks.py", line 289 in __call__
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\runner.py", line 368 in collect_one_node
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\main.py", line 717 in genitems
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\main.py", line 720 in genitems
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\main.py", line 498 in _perform_collect
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\main.py", line 459 in perform_collect
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\main.py", line 266 in pytest_collection
  File "C:\dev\venvs\pywvd\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "C:\dev\venvs\pywvd\lib\site-packages\pluggy\manager.py", line 81 in <lambda>
  File "C:\dev\venvs\pywvd\lib\site-packages\pluggy\manager.py", line 87 in _hookexec
  File "C:\dev\venvs\pywvd\lib\site-packages\pluggy\hooks.py", line 289 in __call__
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\main.py", line 256 in _main
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\main.py", line 213 in wrap_session
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\main.py", line 250 in pytest_cmdline_main
  File "C:\dev\venvs\pywvd\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "C:\dev\venvs\pywvd\lib\site-packages\pluggy\manager.py", line 81 in <lambda>
  File "C:\dev\venvs\pywvd\lib\site-packages\pluggy\manager.py", line 87 in _hookexec
  File "C:\dev\venvs\pywvd\lib\site-packages\pluggy\hooks.py", line 289 in __call__
  File "C:\dev\venvs\pywvd\lib\site-packages\_pytest\config\__init__.py", line 74 in main
  File "C:\dev\venvs\pywvd\lib\site-packages\pytest.py", line 102 in <module>
  File "C:\Users\Gama Info\AppData\Local\Programs\Python\Python37\lib\runpy.py", line 85 in _run_code
  File "C:\Users\Gama Info\AppData\Local\Programs\Python\Python37\lib\runpy.py", line 193 in _run_module_as_main
collected 0 items / 1 errors

======================================================= ERRORS ========================================================
_______________________________________ ERROR collecting tests/test_collect.py ________________________________________
..\venvs\pywvd\lib\site-packages\py\_path\local.py:701: in pyimport
    __import__(modname)
tests\__init__.py:27: in <module>
    for window in desktop:
pyWinVirtualDesktop\__init__.py:407: in __iter__
    id = self.__pDesktopManager.GetWindowDesktopId(hwnd)
E   _ctypes.COMError: (-2147319765, 'Element not found.', (None, None, None, 0, None))
--------------------------------------------------- Captured stdout ---------------------------------------------------
DESKTOP ID: {F7A15DF7-3C90-4327-B702-FCAB45BC7797}
{F7A15DF7-3C90-4327-B702-FCAB45BC7797}
DESKTOP TO LEFT: None
DESKTOP TO RIGHT: None
DESKTOP IS ACTIVE: True
DESKTOP WINDOWS:
--------------------------------------------------- Captured stderr ---------------------------------------------------
Exception ignored in: <function _compointer_base.__del__ at 0x000001DB6013C798>
Traceback (most recent call last):
  File "C:\dev\venvs\pywvd\lib\site-packages\comtypes\__init__.py", line 918, in __del__
    self.Release()
  File "C:\dev\venvs\pywvd\lib\site-packages\comtypes\__init__.py", line 1172, in Release
    return self.__com_Release()
OSError: exception: access violation writing 0x0000000000000000
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=============================================== 1 error in 0.37 seconds ===============================================
```
</details>

I believe the exception is raised while you're trying to access windows not completely filtered after EnumWindows.

In [arrangeit](https://github.com/ipaleka/arrangeit/blob/master/arrangeit/windows/collector.py) I use the following method as filter:
```
    def is_applicable(self, hwnd):
        """Checks if provided hwnd represents window type that should be collected.

        :param hwnd: window id
        :type hwnd: int
        :returns: Boolean
        """
        if not (IsWindow(hwnd) and IsWindowEnabled(hwnd) and IsWindowVisible(hwnd)):
            return False

        if not self._is_alt_tab_applicable(hwnd):
            return False

        if self._is_tray_window(hwnd):
            return False
```

**Comment by ipaleka** [comment]
And so it creates a lot of virtual desktops without destroying them - that's a behavior that should be fixed asap.

**Comment by kdschlosser** [comment]
I shouldn't have to filter the results of EnumWindows right after the Enumeration has finished.

> Enumerates all top-level windows on the screen by passing the handle to each window, in turn, to an application-defined callback function. EnumWindows continues until the last top-level window is enumerated or the callback function returns FALSE.

This is already done internally by the Windows API.

https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-enumwindows

These are the requirements that I know of for GetWindowDesktopId, If there are any others this is something that is going to have to be figured out.

* top level window
* window handle (hwnd) is valid

and these are the reasons why it should not be failing.

* It should not matter if the window is visible or not
* we know the window handles are valid because right before the call to GetWindowDesktopId we called EnumWindows and because windows does a check to ensure a valid window handle we do not have to do this check. If you notice in other places in my code you will see that I do in fact perform this check. This is because a reference of pyWinVirtualDesktop.Window class instance can be held and used later on.


Now if you look here

https://github.com/Ciantic/VirtualDesktopAccessor/blob/master/VirtualDesktopAccessor/dllmain.h

```cpp
GUID DllExport GetWindowDesktopId(HWND window) {
	_RegisterService();

	GUID pDesktopId = {};
	pDesktopManager->GetWindowDesktopId(window, &pDesktopId);

	return pDesktopId;
}
```

You do not see any checking for window handle validity.

I updated my code to add some exception catching during the enumeraing of the windows.

Also. I only create 2 additional desktops during the test. this was because appveyor only had a single desktop and I want to be able to test with more then a single desktop. I have removed the creation of the 2 desktops.

**Comment by kdschlosser** [comment]
It is now passing my appveyor build tests. But I am limited in what I can test because of how old the Windows build is on Appveyor.

It does enumerate the windows properly now,

**Comment by ipaleka** [comment]
Well, probably those 2 desktops are created every time I run tests collecting and raised exception prevented their destroying.

I'll try with the changed code.

**Comment by kdschlosser** [comment]
let me know how you make out. I know it now works with Appveyor. Tho it is limited in what I can do with it.

**Comment by ipaleka** [comment]
I'm not sure how you've predicted to access your classes from the outside, like from the tests. May I rely on `pyWinVirtualDesktop.__original_module__`, its `.__dict__` in particular?

And about testing, as I mentioned, I'm planning to test with the hardcoded arguments against initialized \_methods_\ elements, is it a subclass of `comtypes.IUnknown`, has your enum got assumed values and field names, etc.

You created that Module's \_\_init\_\_ as kind of a smoke test: if that object is successfully initialized it means all your definition and arguments are valid. And as my initial test is falling only on predicted error, then your implementation seems legit... I believe I should test those instances, but dunno if you want them to stay there in \_\_init\_\_ permanently?

**Comment by kdschlosser** [comment]
OK how the setup works. the class `Module` is what I like to call a "dynamic module class" because a module in python is an object just like a class is this is why it works. There are a few attributes that need to be copied over from the "original" module in order to make an exact replica.

By using the code below you replace the module that is imported with the class. There are several reasons to do this.

```python

import sys

class  DynamicModule(object):

    def __init__(self):
        mod = sys.modules[__name__]
        self.__file__ = mod.__file__
        self.__package__ = mod.__package__
        self.__name__ = mod.__name__
        self.__doc__ = mod.__doc__
        self.__path__ = mod.__path__
        self.__orignal_module__ = mod
        sys.modules[__name__] = self

dynamic_module = DynamicModule()
```
by using the dynamic module setup you now have the ability to set up a lazy import mechanism. or a smart import mechanism which ever. you can set that up one of 2 ways. the first way is defining properties for each additional module import and place the import of that module in the property it's self or you can use __getattr__


```python

import sys

class  DynamicModule(object):

    @property
    def some_child_module(self):
        return __import__(__name__ + '.some_child_module`)

    @some_child_module.setter
    def some_child_module(self, value):
        self.__some_child_module = value

    def __getattr__(self, item):
        return __import__(__name__ + '.' + item)

    def __init__(self):
        mod = sys.modules[__name__]
        self.__file__ = mod.__file__
        self.__package__ = mod.__package__
        self.__name__ = mod.__name__
        self.__doc__ = mod.__doc__
        self.__path__ = mod.__path__
        self.__orignal_module__ = mod
        sys.modules[__name__] = self

dynamic_module = DynamicModule()
```

and to use it..

```python
import some_library

some_child_module = some_library.some_child_module

# to use the getattr version
some_other_child_module = some_library.some_other_child_module
```
you would need to have 2 .py files in the same directory as the file that has the DynamicModule class. and those 2 files would need to be named some_child_module.py and some_other_child_module.py

Now you also have the ability to lock out various components. because I did not use these statements in the __init__

* self.__file__ = mod.__file__
* self.__package__ = mod.__package__
* self.__path__ = mod.__path__


If you try running the code below it is going to fail. even tho the module exists.
```python
from pyWinVirtualDesktop import windows_ui
```

I am not sure how your tests work and it could be failing because of something to do with the above.

I set it up this way so that once the module is imported everything is ready to rock and roll there is no need to do anything else. because there is no need to pass any kind of arguments to a constructor and the only reason to import pyWinVirtualDesktop would be if you are running a windows 10 machine there is no reason not to set it up this way. Plus it also allows me to deny access to the other modules in pyWinVirtualDesktop from the program that is importing the library. There is nothing else of use to a user in any of the other files. so there is no need for them to have access to them.

one thing you have to remember when using this kind of a setup. anything that is at the module level that you want the user to have access to must be declared in the __init__ or in a property of the DynamicModule class. otherwise it will not be available to the user.

You can see there are some things I defined at the module level
at the bottom  of the __init__.py file you will see 

```python
pyWinVirtualDesktop = Module()
create_desktop = pyWinVirtualDesktop.create_desktop
```

The only purpose for this is to make an IDE happy. 

technically there is no need to create a reference to the instance of the dynamic module class. there is a reference to it held in sys.modules so it will not get garbage collected. I created a reference to it because i wanted to access it without having to pass it as a parameter. I now have the ability to access it from anywhere inside the file. an example of this is in Desktop.delete()

**Comment by kdschlosser** [comment]
the __original_module__ attribute is there for the purposes of accessing the original module. you can use that and it's dict if you like. in order to import any of the sub modules directly you need to do the following.

```python
import pyWinVirtualDesktop

pyWinVirtualDesktop.__file__ = pyWinVirtualDesktop.__orignal_module__.__file__
pyWinVirtualDesktop.__package__ = pyWinVirtualDesktop.__orignal_module__.__package__
pyWinVirtualDesktop.__name__ = pyWinVirtualDesktop.__orignal_module__.__name__
pyWinVirtualDesktop.__path__ = pyWinVirtualDesktop.__orignal_module__.__path__

from pyWinVirtualDesktop import windows_ui
```

**Comment by kdschlosser** [comment]
and I would prefer to leave the the initialization of IVirtualDesktopManager, IVirtualDesktopManagerInternal,  IServiceProvider, IVirtualDesktopNotificationService, IVirtualDesktopPinnedApps and IApplicationViewCollection in the __init__.. these things do not need to be created over and over again. as they get used. so create the instances a single time and leave it as that. the IVirtualDesktop instances need to be created over and over again because of the possibility of them becoming invalid because the dekstop got deleted. This will also be handled better later on when i get the notifications up and running.

**Comment by ipaleka** [comment]
I'll read everything again tommorow, but basically I at least understand that the current implementation is your choice and I can rely on it while I'll be writting the tests. Not sure now while I'm on smartphone, but iirc I'm not going to need any other data beside that carried with imported classes: for example pyWinVirtualDesktop.\_\_original_module\_\_.IVirtualDesktopManager is everything I need as, for example, IVirtualDesktop that I also need is in the same namespace pyWinVirtualDesktop.\_\_original_module\_\_.

I'll start with the classes definitions and Module.\_\_init\_\_.

**Comment by kdschlosser** [comment]
The reason for my use of the dynamic module is for all intents and purposes Python is considered as a "learning" or "newbie" programmers language. So instead of having to deal with any support issues for someone using something in a manner other then intended I restrict access. a seasoned developer will be able to locate the "hole" to be able to access everything. And those people typically know what they are doing and will not create a support ticket for an issue they are having if doing something outside the scope of the library..

It is simply a way to make accessing components that do not need to be accessed difficult to do. well not difficult but not easy.


and yes you can rely on it. It is going to stay how it is and I have no plans on changing it..

**Comment by ipaleka** [comment]
I made a PR with initial tests.

Now I'm trying to test that `Module.__init__` and I'm not sure how to mock, respectively how to *reset* everything, patch a method, and then reload package. Or just re-instantiate `Module`, whatever...

**Comment by ipaleka** [comment]
I created `requirements.txt` temporary in tests directory, it probably should go in the root.

Are you planning to create a Travis CI account for this repo? I have successfully tested those interfaces with it. Let me know and I'll make a PR with `.travis.yml` configuration file.

**Comment by kdschlosser** [comment]
Why Travis CI? that is geared at Linux isn't it? I have not used it to much and I am not familiar with it.

Take a look at what I just coded up. I decided that it might be better to make a python extension module instead. for several reasons

* Faster
* Easier to obtain information on
* Solves an issue with trying to get IApplicationViewCollection working

[VirtualDesktop.zip](https://github.com/DanEdens/Virtual_Desktops_Plugin/files/3463786/VirtualDesktop.zip)


It's not finished yet and this is just the cpp code for the extension. I still have to write the python side of things as well as the setup routine for compiling it.

**Comment by ipaleka** [comment]
No real reason for Travis CI except it has been the only CI service I've tried to run those Win10 VD related tests on.

**Comment by kdschlosser** [comment]
damned thing. I have a single issue with the extension module and it has to deal with the converting of a GUID into a string to be returned from the extension module. I can get it to not throw errors but the GUID is corrupted and cannot be used for anything. there has to be a way to convert the thing easily.

**Comment by kdschlosser** [comment]
OK so i got that working kind of. it's an in and an out thing. so I got the out working. just not the in. I am going to walk away from it for a day or 2 maybe. that does usually help.

**Comment by ipaleka** [comment]
Yep, that has been the most effective bug discovering technique I know. :)

**Comment by kdschlosser** [comment]
I can't run it locally. and dealing with RDP'n into the appveyor server and having it shutdown every hour is a royal pain in the ass. so debugging is not going to happen. so the best i can do is trial and error. I cannot find an example that actually works they all throw errors. even the examples right on the Microsoft website.. (wow I'm shocked they don't work! NOT)

**Comment by kdschlosser** [comment]
well I finally managed to get that bunch of BS sorted out.
